File: testResampleTensorLayer.R
library( ANTsR )
library( ANTsRNet )
library( magick )
library( tensorflow )
library( keras )


####
#
#  Test 2-D
#

bikeFile <- tempfile( pattern = "bike", fileext = ".jpg" )
download.file(
  url = "https://user-images.githubusercontent.com/22609465/36634042-4168652a-1964-11e8-90a9-2c480b97eff7.jpg",
  destfile = bikeFile )

f <- antsImageRead( bikeFile )
f <- as.array( f )
f <- aperm( f, c( 3, 2, 1 ) )

# ft <- layer_input( shape = dim( f ) )
ft <- layer_input( shape = list( NULL, NULL, 3 ) )

targetShape <- c( 700L, 700L )

types <- c( 'nearestNeighbor', 'linear', 'cubic' )
for( i in seq_len( length( types ) ) )
  {
  cat( "Doing 2-D", types[i], "\n" )
  output <- layer_resample_tensor_2d( ft, targetShape, types[i] )
  model <- keras_model( inputs = ft, outputs = output )
  batchX <- array( data = f, dim = c( 1, dim( f ) ) )
  outputImage <- drop( model %>% predict( batchX ) )

  # Rescale from [0, 1]
  ys <- ( outputImage - min( outputImage ) ) /
        ( max( outputImage ) - min( outputImage ) )
  img <- magick::image_read( ys )
  image_write( img, path =
    paste0( "out_", types[i], format = ".jpg" ) )
  }

####
#
# Brian's variant
#

targetShape <- c( 600L, 700L )

source <- layer_input( shape = list( NULL, NULL, 3 ) )
target <- layer_input( shape = list( NULL, NULL, 1 ) )

sourceImage <- antsImageRead( bikeFile )
targetImage <- as.antsImage( array( data = 0, dim = targetShape ) )

sourceArray <- aperm( as.array( sourceImage ), c( 3, 2, 1 ) )
sourceArray <- array( data = sourceArray, dim = c( 1, dim( sourceArray ) ) )
targetArray <- array( data = as.array( targetImage ), dim = c( 1, dim( targetImage ), 1 ) )

types <- c( 'nearestNeighbor', 'linear', 'cubic' )
for( i in seq_len( length( types ) ) )
  {
  cat( "Doing Brian's 2-D", types[i], "\n" )
  output <- list( source, target ) %>% layer_resample_tensor_to_target_tensor_2d( types[i] )
  model <- keras_model( inputs = list( source, target ), outputs = output )
  outputImage <- drop( model %>% predict( list( sourceArray, targetArray ) ) )

  # Rescale from [0, 1]
  ys <- ( outputImage - min( outputImage ) ) /
        ( max( outputImage ) - min( outputImage ) )
  img <- magick::image_read( ys )
  image_write( img, path =
    paste0( "out_brian_", types[i], format = ".jpg" ) )
  }


####
#
# Test 3-D
#

brainFile <- getANTsRData( "ch2" )

f <- antsImageRead( brainFile )
f <- as.array( f )
f <- array( data = f, dim = c( dim( f ), 1 ) )

# ft <- layer_input( shape = dim( f ) )
ft <- layer_input( shape = list( NULL, NULL, NULL, 1 ) )

shape <- c( 100L, 110L, 120L )

# types <- c( 'nearestNeighbor', 'linear', 'cubic' )
# for( i in seq_len( length( types ) ) )
#   {
#   cat( "Doing 3-D", types[i], "\n" )
#   output <- layer_resample_tensor_3d( ft, shape, types[i] )
#   model <- keras_model( inputs = ft, outputs = output )
#   batchX <- array( data = f, dim = c( 1, dim( f ) ) )
#   outputArray <- model %>% predict( batchX, verbose = TRUE )

#   # we skip calculating the correct header info as it's not
#   # important for demonstrating functionality.
#   outputImage <- as.antsImage( drop( outputArray ) )
#   antsImageWrite( outputImage, paste0( "out3D_", types[i], ".nii.gz" ) )
#   }


####
#
# Brian's variant
#

targetShape <- c( 100L, 110L, 120L )

source <- layer_input( shape = list( NULL, NULL, NULL, 1 ) )
target <- layer_input( shape = list( NULL, NULL, NULL, 1 ) )

sourceArray <- array( data = f, dim = c( 1, dim( f ) ) )
targetArray <- array( data = 0, dim = c( 1, targetShape, 1 ) )

types <- c( 'nearestNeighbor', 'linear', 'cubic' )
for( i in seq_len( length( types ) ) )
  {
  cat( "Doing Brian's 3-D", types[i], "\n" )
  output <- list( source, target ) %>% layer_resample_tensor_to_target_tensor_3d( types[i] )
  model <- keras_model( inputs = list( source, target ), outputs = output )
  outputArray <- model %>% predict( list( sourceArray, targetArray ) )

  # we skip calculating the correct header info as it's not
  # important for demonstrating functionality.
  outputImage <- as.antsImage( drop( outputArray ) )
  antsImageWrite( outputImage, paste0( "out3D_brian_", types[i], ".nii.gz" ) )
  }

File: testResampleTensorLayer.py
import requests
import tempfile
import numpy as np
from tensorflow.keras import Model
from tensorflow.keras.layers import Input
from os import path
import cv2

import ants
import antspynet


####
#
#  Test 2-D
#

url = "https://user-images.githubusercontent.com/22609465/36634042-4168652a-1964-11e8-90a9-2c480b97eff7.jpg"
r = requests.get(url)

temp_directory = tempfile.TemporaryDirectory()
target_file = tempfile.NamedTemporaryFile(suffix=".jpg", dir=temp_directory.name)
target_file.close()
target_file_name = target_file.name

if not path.exists(target_file_name):
    r = requests.get(url)
    with open(target_file_name, 'wb') as f:
        f.write(r.content)

f = cv2.imread(target_file_name)

# ft = Input(shape=f.shape)
ft = Input(shape=(None, None, 3))

shape = (700, 700)

types = ['nearest_neighbor', 'linear', 'cubic']
for i in range(len(types)):
    output = antspynet.ResampleTensorLayer2D(shape, types[i])(ft)
    model = Model(inputs=ft, outputs=output)
    batchX = np.expand_dims(f, axis = 0)
    output_image = model.predict(batchX)
    cv2.imwrite("out2D_" + types[i] + "_py.jpg", output_image[0])

####
#
#  Brian's variant
#

ft = Input(shape=(None, None, 3))

target_shape = (600, 700)

source = Input(shape = (None, None, 3))
target = Input(shape = (None, None, 3))

source_array = np.expand_dims(f, axis = 0)
target_array = np.zeros((1, *target_shape, 3))

types = ['nearest_neighbor', 'linear', 'cubic']
for i in range(len(types)):
    print("Doing Brian's 2-D " + types[i])
    output = antspynet.ResampleTensorToTargetTensorLayer2D(types[i])([source, target])
    model = Model(inputs=[source, target], outputs=output)
    batchX = np.expand_dims(f, axis = 0)
    output_image = model.predict([source_array, target_array])
    cv2.imwrite("out2D_brian_" + types[i] + "_py.jpg", output_image[0])

####
#
# Test 3-D
#

brain_file = ants.get_ants_data("ch2")

f = ants.image_read(brain_file)
f = np.expand_dims( f.numpy(), axis=-1)

shape = (100, 110, 120)
ft = Input(shape=(None, None, None, 1))
# ft = Input(shape=f.shape)

types = ['nearest_neighbor', 'linear', 'cubic']
for i in range(len(types)):
    output = antspynet.ResampleTensorLayer3D(shape, types[i])(ft)
    model = Model(inputs=ft, outputs=output)
    batchX = np.expand_dims(f, axis = 0)
    output_image = ants.from_numpy(np.squeeze(model.predict(batchX)))
    ants.image_write(output_image, "out3D_py_" + types[i] + ".nii.gz")

####
#
#  Brian's variant
#

target_shape = (100, 110, 120)

source = Input(shape = (None, None, None, 1))
target = Input(shape = (None, None, None, 1))

source_array = np.expand_dims(f, axis = 0)
target_array = np.zeros((1, *target_shape, 1))

types = ['nearest_neighbor', 'linear', 'cubic']
for i in range(len(types)):
    print("Doing Brian's 3-D " + types[i])
    output = antspynet.ResampleTensorToTargetTensorLayer3D(types[i])([source, target])
    model = Model(inputs=[source, target], outputs=output)
    output_image = ants.from_numpy(np.squeeze(model.predict([source_array, target_array])))
    ants.image_write(output_image, "out3D_brian_py_" + types[i] + ".nii.gz")

